repose-aggregator/components/services/datastore-service/datastore-service-impl/src/test/scala/org/openrepose/nodeservice/distributed/servlet/DistributedDatastoreServletTest.scala

```diff
@@ -0,0 +1,93 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.nodeservice.distributed.servlet
+
+import java.net.InetAddress
+import java.util.Collections
+import javax.servlet.http.HttpServletResponse
+import javax.servlet.http.HttpServletResponse._
+
+import io.opentracing.Tracer
+import io.opentracing.mock.MockTracer
+import org.junit.runner.RunWith
+import org.mockito.Matchers.anyString
+import org.mockito.Mockito
+import org.mockito.Mockito.verify
+import org.openrepose.core.services.datastore.distributed.ClusterConfiguration
+import org.openrepose.core.services.datastore.distributed.config._
+import org.openrepose.core.services.datastore.impl.distributed.CacheRequest.CACHE_URI_PATH
+import org.openrepose.core.services.datastore.{Datastore, DatastoreAccessControl, DatastoreService}
+import org.scalatest.junit.JUnitRunner
+import org.scalatest.mock.MockitoSugar
+import org.scalatest.{BeforeAndAfterEach, FunSpec, Matchers}
+import org.springframework.mock.web.{MockHttpServletRequest, MockHttpServletResponse}
+
+
+@RunWith(classOf[JUnitRunner])
+class DistributedDatastoreServletTest extends FunSpec with BeforeAndAfterEach with Matchers with MockitoSugar {
+
+  private val mockDatastoreService = mock[DatastoreService]
+  private val mockDatastore = mock[Datastore]
+  Mockito.when(mockDatastoreService.getDefaultDatastore).thenReturn(mockDatastore)
+  private val distributedDatastoreConfiguration = mock[DistributedDatastoreConfiguration]
+  val tracerSpy = Mockito.spy(new MockTracer)
+
+  val distributedDatastoreServlet = new DistributedDatastoreServlet(
+    mockDatastoreService,
+    mock[ClusterConfiguration],
+    new DatastoreAccessControl(Collections.emptyList[InetAddress], true),
+    distributedDatastoreConfiguration,
+    tracerSpy
+  )
+
+  var servletRequest: MockHttpServletRequest = _
+  var servletResponse: HttpServletResponse = _
+
+
+  override def beforeEach(): Unit = {
+    servletRequest = new MockHttpServletRequest
+    servletRequest.setRequestURI(CACHE_URI_PATH)
+    servletRequest.setProtocol("1.1")
+    servletResponse = new MockHttpServletResponse
+    tracerSpy.reset()
+  }
+
+  describe("Distributed Datastore calls without a Cache Key") {
+    val notFound = List("GET", "HEAD").map(method => (method, SC_NOT_FOUND)) // Head calls Get and no Cache Key sent
+    val badRequest = List("PUT", "PATCH").map(method => (method, SC_BAD_REQUEST)) // No data or Cache Key sent
+    val notAllowed = List("POST", "TRACE").map(method => (method, SC_METHOD_NOT_ALLOWED)) // Methods Not Allowed
+    val okMethods = List("OPTIONS").map(method => (method, SC_OK)) // This is handled by the parent HTTP Servlet
+    val notImplemented = List("CUSTOM", "BOGUS").map(method => (method, SC_NOT_IMPLEMENTED)) // Not Implemented
```

Comment:
This looks an awful lot like a map to me. You could either go for the straightforward method-to-response-code map like:
```
val methodToResponseCode = Map(
  "GET" -> SC_NOT_FOUND,
  "HEAD" -> SC_NOT_FOUND,
  "PUT" -> SC_BAD_REQUEST,
  ...)

methodToResponseCode foreach { case (httpMethod, responseCode) =>
  ...
}
```
or if you want to maintain your groupings and perhaps the comments:
```
val responseCodeToMethods = Map(
  SC_NOT_FOUND -> Set("GET", "HEAD"), // Head calls Get and no Cache Key sent
  SC_BAD_REQUEST -> Set("PUT", "PATCH") // No data or Cache Key sent
  ...)

responseCodeToMethods foreach { case (responseCode, methods) =>
  methods foreach { method =>
    ...
  }
}
```

Reply:
You are correct, it could be a `Map`. I just chose to break it out into a `Seq` of tuples. Converting it to a single `Map` with redundant codes explicitly coded just seemed, well redundant. Converting it to a `Map` of `Seq`'s seems overly complex and causes more code in the end with the need of the extra `foreach`.

I swapped the `List` for the more generic `Seq` and moved to the less verbose underscore notation for the `map` method. I did leave the `okMethods` creation in the more verbose form for consistency even though it could have simply been defined as a single tuple.

Reply:
To be fair, I believe the `Map` of `Seq`s is still less code (since you lose all the `.map(...)` calls), but it does add another scope level. To get around that, you could do something like:
```
val responseCodeToMethods = Map(
  SC_NOT_FOUND -> Set("GET", "HEAD"), // Head calls Get and no Cache Key sent
  SC_BAD_REQUEST -> Set("PUT", "PATCH") // No data or Cache Key sent
  ...)

responseCodeToMethods.toSeq.flatMap({case (key, values) => values.map(key -> _)}) foreach { case (responseCode, method) =>
  ...
}
```

A `Map` still seems more natural to me, but whether your data structure is a `Seq` of tuples or a `Map`, you have to do some transformation. For the `Seq`, first off, you have to declare many independendent `Seq`s and combine them. For each `Seq`, you have to `.map(...)` all of the values into a tuple. For the `Map`, you have to (at least, to drop the nested `.foreach(...)`) break the one-key-to-value constraint. Or you accept the additional nested scope.

I guess I don't care too much, it just seemed to me like you should call a map a map.

Reply:
The `Map` is a `Map`.